### PR TITLE
feat(ui): Batch set & unset Domain for assets via the UI

### PR DESCRIPTION
**Pull request by jjoyce0510**
# Summary

In this PR, we introduce support for applying Domains in bulk to assets via the UI. 

# Sceenshots
Screen Shot 2022-08-03 at 2 15 24 PM
Screen Shot 2022-08-03 at 2 15 33 PM

## Checklist
- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/mutate/BatchSetDomainResolver.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+  private void validateInputResources(List<ResourceRefInput> resources, QueryContext context) {
+    for (ResourceRefInput resource : resources) {
+      validateInputResource(resource, context);
+    }
+  }
+
+  private void validateInputResource(ResourceRefInput resource, QueryContext context) {
+    final Urn resourceUrn = UrnUtils.getUrn(resource.getResourceUrn());
+    if (!LabelUtils.isAuthorizedToUpdateTerms(context, resourceUrn, resource.getSubResource())) {
```
Comment (chriscollins3456) [inline datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/mutate/BatchSetDomainResolver.java:68]: should be `isAuthorizedToUpdateDomainsForEntity` right?

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/mutate/util/DomainUtils.java [hunk 2]
```diff
@@ -33,4 +45,49 @@ public static boolean isAuthorizedToUpdateDomainsForEntity(@Nonnull QueryContext
...
+    Domains domains = (Domains) getAspectFromEntity(
+        resource.getResourceUrn(),
+        Constants.DOMAINS_ASPECT_NAME,
+        entityService,
+        new Domains());
+    final UrnArray newDomains = new UrnArray();
+    if (domainUrn != null) {
+      newDomains.add(domainUrn);
+    }
+    domains.setDomains(newDomains);
```
Comment (chriscollins3456) [inline datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/mutate/util/DomainUtils.java:77]: do we want to allow users to set multiple domains on an asset?

File: datahub-graphql-core/src/test/java/com/linkedin/datahub/graphql/resolvers/domain/BatchSetDomainResolverTest.java [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+    Mockito.when(mockService.exists(Urn.createFromString(TEST_DOMAIN_2_URN))).thenReturn(true);
+
+    BatchSetDomainResolver resolver = new BatchSetDomainResolver(mockService);
+
+    // Execute resolver
+    QueryContext mockContext = getMockAllowContext();
+    DataFetchingEnvironment mockEnv = Mockito.mock(DataFetchingEnvironment.class);
+    BatchSetDomainInput input = new BatchSetDomainInput(TEST_DOMAIN_2_URN, ImmutableList.of(
+        new ResourceRefInput(TEST_ENTITY_URN_1, null, null),
+        new ResourceRefInput(TEST_ENTITY_URN_2, null, null)));
```
Comment (chriscollins3456) [inline datahub-graphql-core/src/test/java/com/linkedin/datahub/graphql/resolvers/domain/BatchSetDomainResolverTest.java:134]: relevant test to previous question on if we should allow them to add a domain to an asset that already has a domain

File: datahub-web-react/src/app/entity/shared/components/styled/search/action/DomainsDropdown.tsx [hunk 4]
```diff
@@ -1,27 +1,78 @@
...
+                variables: {
+                    input: {
+                        resources: [...urns.map((urn) => ({ resourceUrn: urn }))],
+                    },
                 },
-            ]}
-            disabled={disabled}
-        />
+            });
+            message.success({ content: 'Removed Domain!', duration: 2 });
```
Comment (chriscollins3456) [inline datahub-web-react/src/app/entity/shared/components/styled/search/action/DomainsDropdown.tsx:27]: not that it matters much and I know this is how it's been in your other batch PRs, but is there a particular reason we are using an `async` function and awaiting the result, as opposed to using the promise`.then(() => {} ).catch(() => {})` way that we do most other places?

File: datahub-web-react/src/app/entity/shared/containers/profile/sidebar/Domain/SetDomainModal.tsx [hunk 5]
```diff
@@ -30,15 +30,14 @@ const StyleTag = styled(Tag)`
     align-items: center;
 `;
 
-export const SetDomainModal = ({ onCloseModal, refetch }: Props) => {
+export const SetDomainModal = ({ urns, onCloseModal, refetch }: Props) => {
     const entityRegistry = useEntityRegistry();
-    const urn = useMutationUrn();
```
Comment (chriscollins3456) [inline datahub-web-react/src/app/entity/shared/containers/profile/sidebar/Domain/SetDomainModal.tsx:35]: what happens if the user selects a sibling entity from the multi-select option? Will we know to use the correct urn (ie. the primary sibling)?

**Review by chriscollins3456 — COMMENTED** [review]
Nice! I have a few questions for ya, other than that good to go

**Comment by xiphl** [comment]
Have the team thought about considered allowing multiple domains to a dataset in UI yet? (I know backend supports it)

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/mutate/BatchSetDomainResolver.java (hunk 1 above)
Comment (jjoyce0510) [reply datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/mutate/BatchSetDomainResolver.java:68]: O gosh thank you

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/mutate/util/DomainUtils.java (hunk 2 above)
Comment (jjoyce0510) [reply datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/mutate/util/DomainUtils.java:77]: Currently the product only allows one. The underlying data model was designed to support extension should this be a high-demand ask

File: datahub-graphql-core/src/test/java/com/linkedin/datahub/graphql/resolvers/domain/BatchSetDomainResolverTest.java (hunk 3 above)
Comment (jjoyce0510) [reply datahub-graphql-core/src/test/java/com/linkedin/datahub/graphql/resolvers/domain/BatchSetDomainResolverTest.java:134]: Yeah so you can change (or unset a domain)

File: datahub-web-react/src/app/entity/shared/components/styled/search/action/DomainsDropdown.tsx (hunk 4 above)
Comment (jjoyce0510) [reply datahub-web-react/src/app/entity/shared/components/styled/search/action/DomainsDropdown.tsx:27]: No we can change it to keep consistency, I think this is copied from another place we use

File: datahub-web-react/src/app/entity/shared/containers/profile/sidebar/Domain/SetDomainModal.tsx (hunk 5 above)
Comment (jjoyce0510) [reply datahub-web-react/src/app/entity/shared/containers/profile/sidebar/Domain/SetDomainModal.tsx:35]: That's a great question. Right now we merge down multiple results on the same page to a single entity, with the primary sibling showing in the search results. So it should go back to the primary

**Comment by jjoyce0510** [comment]
@xiphl we have. But from a semantics PoV, we feel that restricting to a single domain still offers an alternative to other mechanisms for categorizing data that we have like tags and terms... I'm personally concerned that if we support multiple domains people will begin to question why we have both glossary and domains, and what the practical difference is..

File: datahub-web-react/src/app/entity/shared/containers/profile/sidebar/Domain/SetDomainModal.tsx (hunk 5 above)
Comment (jjoyce0510) [reply datahub-web-react/src/app/entity/shared/containers/profile/sidebar/Domain/SetDomainModal.tsx:35]: This type of hook should not be in a general purpose modal regardless, as it assumes that its being invoked from an entity profile

File: datahub-web-react/src/app/entity/shared/components/styled/search/action/DomainsDropdown.tsx (hunk 4 above)
Comment (jjoyce0510) [reply datahub-web-react/src/app/entity/shared/components/styled/search/action/DomainsDropdown.tsx:27]: updating!

File: datahub-web-react/src/app/entity/shared/containers/profile/sidebar/Domain/SetDomainModal.tsx (hunk 5 above)
Comment (chriscollins3456) [reply datahub-web-react/src/app/entity/shared/containers/profile/sidebar/Domain/SetDomainModal.tsx:35]: makes sense! yeah, understood why you took it out of here and wanted to pass an urn in, just wanted to make sure the search modal had you select the primary entity

**Review by chriscollins3456 — APPROVED** [review]
cool beans

**Pull request merged**